### PR TITLE
Niftireader

### PR DESCRIPTION
**Pull request by ssktotoro**
## Before submitting (checklist)

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)?
- [x] Did you check the code style? `catalyst-make-codestyle && catalyst-check-codestyle` (`pip install -U catalyst-codestyle`).
- [x] Did you make sure to update the docs? We use Google format for all the methods and classes.
- [x] Did you check the docs with `make check-docs`?
- [x] Did you write any new necessary tests?
- [x] Did you check that your code passes the unit tests `pytest .` ? 
- [x] Did you add your new functionality to the docs?
- [x] Did you update the [CHANGELOG](https://github.com/catalyst-team/catalyst/blob/master/CHANGELOG.md)? No I need to make the PR to link first
- [ ] Did you run [colab minimal CI/CD](https://colab.research.google.com/drive/1JCGTVvWlrIsLXMPRRRSWiAstSLic4nbA) with `latest` and `minimal` requirements?

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

PS
- [x] I know, that I could [join slack](https://join.slack.com/t/catalyst-team-core/shared_invite/zt-d9miirnn-z86oKDzFMKlMG4fgFdZafw) for pull request discussion.

### CODE REVIEW

**Comment by Scitator** [comment]
Thanks for the PR! Could you please also:
- add a new class to the docs, like https://github.com/catalyst-team/catalyst/blob/master/docs/api/data.rst#readers
- add new requirements to github CI [codestyle](https://github.com/catalyst-team/catalyst/blob/master/.github/workflows/codestyle.yml) and [checks](https://github.com/catalyst-team/catalyst/blob/master/.github/workflows/dl_cpu.yml)
- as far as you have `>=` requirements we should also add it to the minimal requirements [here](https://github.com/catalyst-team/catalyst/blob/master/.github/workflows/dl_cpu.yml)

PS. would you like to contribute any extra test for NiftyReader for the community?

**Comment by ssktotoro** [comment]
is there a test for existing readers like the cv reader?

**Comment by Scitator** [comment]
Not sure if :)
(our main contrib issue)

But you could use our [utils](https://github.com/catalyst-team/catalyst/tree/master/catalyst/utils/tests) or [data](https://github.com/catalyst-team/catalyst/tree/master/catalyst/data/tests) test for inspiration.

**Comment by ssktotoro** [comment]
@Scitator not really sure how to fix the cache-pip issue >.<
I don't know how the .github/workflows/codestyle.yml works

File: catalyst/contrib/data/nifti/reader.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
+from typing import Optional
+
+import nibabel as nib
```
Comment (Scitator) [inline catalyst/contrib/data/nifti/reader.py]:
```suggestion
from catalyst.settings import SETTINGS
if SETTINGS.nifti_required:
    import nibabel as nib
```

by this way, we could build the documentation even without the `nibabel` package :)

**Comment by Scitator** [comment]
Looks like you have change the [reviewdog workflow](https://github.com/catalyst-team/catalyst/blob/3260296407586834228e5fe01a082703c2ae6398/.github/workflows/codestyle.yml#L16#L70). Nevertheless, we also need to add requirements update: cache and installation to the [codestyle-and-docs one](https://github.com/catalyst-team/catalyst/blob/3260296407586834228e5fe01a082703c2ae6398/.github/workflows/codestyle.yml#L72#L119). That's the case :)

File: .github/workflows/codestyle.yml [hunk 2]
```diff
@@ -45,14 +45,22 @@ jobs:
...
+          key: ${{ runner.os }}-${{ matrix.python-version }}-pip -${{ hashFiles('./requirements/requirements.txt') }} -${{ hashFiles('./requirements/requirements-dev.txt') }}
+          restore-keys: |
+            ${{ runner.os }}-${{ matrix.python-version }}-pip-
+
+      - name: cache pip contrib
+        uses: actions/cache@v2
+        with:
+          path: ${{ steps.pip-cache.outputs.dir }}
+          key: ${{ runner.os }}-${{ matrix.python-version }}-pip -${{ hashFiles('./requirements/requirements-cv.txt') }} -${{ hashFiles('./requirements/requirements-hydra.txt') }} -${{ hashFiles('./requirements/requirements-ml.txt') }} -${{ hashFiles('./requirements/requirements-nifti.txt') }} -${{ hashFiles('./requirements/requirements-optuna.txt') }} -${{ hashFiles('./requirements/requirements-mlflow.txt') }}
```
Comment (Scitator) [inline .github/workflows/codestyle.yml]: I think we need to change the `key` in case of several caches

File: .github/workflows/codestyle.yml [hunk 3]
```diff
@@ -99,7 +107,15 @@ jobs:
...
```
Comment (Scitator) [inline .github/workflows/codestyle.yml]: and here

**Review by Scitator — DISMISSED** [review]

File: .github/workflows/codestyle.yml (hunk 2 above)
Comment (ssktotoro) [reply .github/workflows/codestyle.yml]: can you point me to either what consumes codestyle.yml so I know what is valid/ good for the yml?

Or you can suggest directly what I'm supposed to change.  I don't know what consumes the yml and what the key even means and therefore don't know what to change the key to...

File: .github/workflows/codestyle.yml (hunk 3 above)
Comment (ssktotoro) [reply .github/workflows/codestyle.yml]: see above.. I don't understand the workflow/ use of this codestyle.yml and therefore can't suggest a better key.

File: .github/workflows/codestyle.yml (hunk 2 above)
Comment (Scitator) [reply .github/workflows/codestyle.yml]: honestly speaking, I suggest just return the initial case with united pip cache
I need an extra time to check GitHub CI docs, so we could update the cache logic later

Comment (ssktotoro) [reply .github/workflows/codestyle.yml]: I can do that... just fyi the test will fail b/c the key in the initial case is too long.

Comment (Scitator) [reply .github/workflows/codestyle.yml]: oh

Comment (Scitator) [inline .github/workflows/codestyle.yml]:
```suggestion
          key: ${{ runner.os }}-${{ matrix.python-version }}-pip-contrib -${{ hashFiles('./requirements/requirements-cv.txt') }} -${{ hashFiles('./requirements/requirements-hydra.txt') }} -${{ hashFiles('./requirements/requirements-ml.txt') }} -${{ hashFiles('./requirements/requirements-nifti.txt') }} -${{ hashFiles('./requirements/requirements-optuna.txt') }} -${{ hashFiles('./requirements/requirements-mlflow.txt') }}
```

File: .github/workflows/codestyle.yml (hunk 3 above)
Comment (Scitator) [inline .github/workflows/codestyle.yml]:
```suggestion
          key: ${{ runner.os }}-${{ matrix.python-version }}-pip-contrib -${{ hashFiles('./requirements/requirements-cv.txt') }} -${{ hashFiles('./requirements/requirements-hydra.txt') }} -${{ hashFiles('./requirements/requirements-ml.txt') }} -${{ hashFiles('./requirements/requirements-nifti.txt') }} -${{ hashFiles('./requirements/requirements-optuna.txt') }} -${{ hashFiles('./requirements/requirements-mlflow.txt') }}
```

**Review by Scitator — COMMENTED** [review]
could we do so?

File: .github/workflows/codestyle.yml (hunk 3 above)
Comment (ssktotoro) [reply .github/workflows/codestyle.yml]: yup let's do it. Just finishing up the minimal test.

**Comment by ssktotoro** [comment]
The test makes CI fail as the requirements for contrib are presumably not installed.  Not sure what to do about it.

File: .github/workflows/codestyle.yml [hunk 4]
```diff
@@ -45,14 +45,22 @@ jobs:
...
+          restore-keys: |
+            ${{ runner.os }}-${{ matrix.python-version }}-pip-
+
+      - name: cache pip contrib
+        uses: actions/cache@v2
+        with:
+          path: ${{ steps.pip-cache.outputs.dir }}
+          key: ${{ runner.os }}-${{ matrix.python-version }}-pip-contrib -${{ hashFiles('./requirements/requirements-cv.txt') }} -${{ hashFiles('./requirements/requirements-hydra.txt') }} -${{ hashFiles('./requirements/requirements-ml.txt') }} -${{ hashFiles('./requirements/requirements-nifti.txt') }} -${{ hashFiles('./requirements/requirements-optuna.txt') }} -${{ hashFiles('./requirements/requirements-mlflow.txt') }}
           restore-keys: |
             ${{ runner.os }}-${{ matrix.python-version }}-pip-
```
Comment (Scitator) [inline .github/workflows/codestyle.yml]:
```suggestion
            ${{ runner.os }}-${{ matrix.python-version }}-pip-contrib
```

File: .github/workflows/codestyle.yml [hunk 5]
```diff
@@ -99,7 +107,15 @@ jobs:
...
```
Comment (Scitator) [inline .github/workflows/codestyle.yml]:
```suggestion
            ${{ runner.os }}-${{ matrix.python-version }}-pip-contrib
```

**Review by Scitator — COMMENTED** [review]
could that help?

File: .github/workflows/codestyle.yml (hunk 5 above)
Comment (ssktotoro) [reply .github/workflows/codestyle.yml]: apparently not...

Comment (ssktotoro) [reply .github/workflows/codestyle.yml]: ah one sec...

Comment (ssktotoro) [reply .github/workflows/codestyle.yml]: I needed to change the push file...

**Comment by ssktotoro** [comment]
don't know why catalyst / dl-cpu (ubuntu-20.04, 3.7, minimal) (pull_request)  fails, but it should be good?

**Comment by Scitator** [comment]
that should be okay – Github CI freezes sometimes

**Pull request merged**